File: labs/control-ui/api-shadow/cdn_front_pool.ts
export interface CdnCandidate {
  ip: string;
  sni: string;
  port: number;
  rttEmaMs: number;
  successCount: number;
  failureCount: number;
  quarantinedUntilMs: number;
}

export class CdnFrontPool {
  private candidates: CdnCandidate[] = [];
  private readonly quarantineDurationMs: number;
  private readonly maxCapacity: number;
  constructor(quarantineDurationMs: number = 60000, maxCapacity: number = 100) {
    this.quarantineDurationMs = quarantineDurationMs;
    this.maxCapacity = maxCapacity;
  }

  public addCandidate(ip: string, sni: string, port = 443): boolean {
    if (this.candidates.some((c) => c.ip === ip && c.sni === sni && c.port === port)) {
      return true;
    }
    if (this.candidates.length >= this.maxCapacity) {
      return false;
    }
    this.candidates.push({
      ip,
      sni,
      port,
      rttEmaMs: 0,
      successCount: 0,
      failureCount: 0,
      quarantinedUntilMs: 0,
    });
    return true;
  }

  public recordSuccess(ip: string, rttMs: number): void {
    const cand = this.candidates.find((c) => c.ip === ip);
    if (!cand) return;
    cand.successCount++;
    cand.rttEmaMs = cand.rttEmaMs === 0 ? rttMs : 0.8 * cand.rttEmaMs + 0.2 * rttMs;
  }

  public recordFailure(ip: string, nowMs = Date.now()): void {
    const cand = this.candidates.find((c) => c.ip === ip);
    if (!cand) return;
    cand.failureCount++;
    if (cand.failureCount % 3 === 0) {
      cand.quarantinedUntilMs = nowMs + this.quarantineDurationMs;
    }
  }

  public selectBest(nowMs = Date.now()): CdnCandidate | null {
    const healthy = this.candidates.filter((c) => nowMs >= c.quarantinedUntilMs);
    if (healthy.length === 0) return null;
    return healthy.reduce((best, cur) => {
      const scoreBest = best.rttEmaMs === 0 ? 50 : best.rttEmaMs;
      const scoreCur = cur.rttEmaMs === 0 ? 50 : cur.rttEmaMs;
      return scoreCur < scoreBest ? cur : best;
    });
  }

  public getCandidates(): readonly CdnCandidate[] {
    return this.candidates;
  }
}

File: labs/mobile/donor-archive/tunnel/CdnFrontPool.kt
// Copyright 2024-2026 LumiNet Authors. All Rights Reserved.
package com.luminet.android.tunnel

data class CdnCandidate(
    val ip: String,
    val sni: String,
    val port: Int = 443,
    var rttEmaMs: Double = 0.0,
    var successCount: Long = 0,
    var failureCount: Long = 0,
    var quarantinedUntilMs: Long = 0
) {
    fun isQuarantined(nowMs: Long = System.currentTimeMillis()): Boolean = nowMs < quarantinedUntilMs

    fun recordSuccess(rttMs: Double) {
        successCount++
        rttEmaMs = if (rttEmaMs == 0.0) rttMs else 0.8 * rttEmaMs + 0.2 * rttMs
    }

    fun recordFailure(quarantineDurationMs: Long = 60_000, nowMs: Long = System.currentTimeMillis()) {
        failureCount++
        if (failureCount % 3L == 0L) {
            quarantinedUntilMs = nowMs + quarantineDurationMs
        }
    }
}

class CdnFrontPool(
    private val quarantineDurationMs: Long = 60_000,
    private val maxCapacity: Int = 100
) {
    private val candidates = mutableListOf<CdnCandidate>()
    private val lock = Any()

    fun addCandidate(ip: String, sni: String, port: Int = 443): Boolean = synchronized(lock) {
        if (candidates.any { it.ip == ip && it.sni == sni && it.port == port }) return true
        if (candidates.size >= maxCapacity) return false
        candidates.add(CdnCandidate(ip, sni, port))
        true
    }

    fun selectBest(nowMs: Long = System.currentTimeMillis()): CdnCandidate? = synchronized(lock) {
        candidates.filter { !it.isQuarantined(nowMs) }
            .minByOrNull { if (it.rttEmaMs == 0.0) 50.0 else it.rttEmaMs }
    }
}
